### PR TITLE
cloud/cloudimpl: adds KMS API and AWS KMS integration

### PR DESCRIPTION
**Pull request by adityamaru**
This change introduces a KMS API and a concrete implementation of the
AWS KMS service. It also adds a simple encrypt-decrypt unit test.

This is v2 of the API, discussions about v1 can be found at #51779.

Release note: None

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -24,7 +24,7 @@ require (
 	github.com/apache/arrow/go/arrow v0.0.0-20200610220642-670890229854
 	github.com/apache/thrift v0.0.0-20181211084444-2b7365c54f82 // indirect
 	github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e
-	github.com/aws/aws-sdk-go v1.16.19
+	github.com/aws/aws-sdk-go v1.33.8
```
Comment (adityamaru) [inline go.mod:27]: I followed this - https://github.com/cockroachdb/cockroach/blob/master/build/README.md#dependencies and pushed a named branch to cockroachdb/vendored. Should the go.mod/go.sum changes not show up here?

**Review by miretskiy — APPROVED** [review]
> Reviewed 11 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[go.mod, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51884#-MD13sk5GXZZA88GQQlT-r1-27:-MD1BUfJ8_VMnvcL2f06:bcgb0wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/266867053f6b13947a3d2dfe40b03b5fa4286ad4/go.mod#L27)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I followed this - https://github.com/cockroachdb/cockroach/blob/master/build/README.md#dependencies and pushed a named branch to cockroachdb/vendored. Should the go.mod/go.sum changes not show up here?
</blockquote></details>

I think that pr should be submitted/merged & then you rebase on top of that...
___
*[pkg/storage/cloudimpl/cloudimpltests/aws_kms_test.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51884#-MD1Bxjr5iy71B_H3nYO:-MD1Bxjr5iy71B_H3nYP:b-iq0io8) ([raw file](https://github.com/cockroachdb/cockroach/blob/266867053f6b13947a3d2dfe40b03b5fa4286ad4/pkg/storage/cloudimpl/cloudimpltests/aws_kms_test.go#L68)):*
> ```Go
> 
> 	t.Run("auth-empty-no-cred", func(t *testing.T) {
> ```

extra \n?

File: pkg/storage/cloudimpl/aws_kms.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	// Luckily, AWS secrets are base64-encoded data and thus will never actually
+	// contain spaces. We can convert any space characters we see to +
+	// characters to recover the original secret.
+	kmsURIParams.secret = strings.Replace(kmsURIParams.secret, " ", "+", -1)
+	return kmsURIParams
+}
+
+// MakeKMSFromURI is the factory method which returns a configured, ready-to-use
+// AWS KMS object.
+func MakeKMSFromURI(uri string, env cloud.KMSEnv) (cloud.KMS, error) {
```
Comment (dt) [inline pkg/storage/cloudimpl/aws_kms.go:70]: Should this be `MakeAWSKMS` to indicate its AWS-speicific nature? Also, if all (one of) the constructors are from URI, we can probably skip "FromURI" suffix.

File: pkg/storage/cloudimpl/aws_kms.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	secret    string
+	tempToken string
+	endpoint  string
+	region    string
+	auth      string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	kmsURIParams := kmsURIParams{
+		kmsURI.Query().Get(AWSAccessKeyParam),
```
Comment (dt) [inline pkg/storage/cloudimpl/aws_kms.go:49]: I've never said "I'm glad I skipped those struct field names" but I sure have said the opposite.

File: go.mod (hunk 1 above)
Comment (dt) [reply go.mod:27]: I believe these changes are correct: these changes show up here, and are consistent with the content in your updated `vendor` ref I assume. When this PR merges it'll bump the submodule pointer to your ref in vendored, and these manifest changes will merge so the manifest remains consistent with what is checked out of the submodule.

File: pkg/storage/cloudimpl/aws_kms.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	accessKey string
+	secret    string
+	tempToken string
+	endpoint  string
+	region    string
+	auth      string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	kmsURIParams := kmsURIParams{
```
Comment (dt) [inline pkg/storage/cloudimpl/aws_kms.go:48]: shadowing always leads to regret. I'd call it `params` or `p`

**Review by dt — APPROVED** [review]

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[pkg/storage/cloudimpl/aws_kms.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51884#-MD13sk5GXZZA88GQQlY-r2-48:-MDFUvirEL1unTGAvElu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a18377d9277cce50967226e12afc1690d782908/pkg/storage/cloudimpl/aws_kms.go#L48)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

shadowing always leads to regret. I'd call it `params` or `p`
</blockquote></details>

Done.
___
*[pkg/storage/cloudimpl/aws_kms.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51884#-MD13sk5GXZZA88GQQlY-r2-49:-MDFUw0i3pG41ncMRnlk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a18377d9277cce50967226e12afc1690d782908/pkg/storage/cloudimpl/aws_kms.go#L49)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I've never said "I'm glad I skipped those struct field names" but I sure have said the opposite.
</blockquote></details>

Done.
___
*[pkg/storage/cloudimpl/aws_kms.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51884#-MD13sk5GXZZA88GQQlY-r2-70:-MDFUdmA2rnIHUvGtoQm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a18377d9277cce50967226e12afc1690d782908/pkg/storage/cloudimpl/aws_kms.go#L70)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Should this be `MakeAWSKMS` to indicate its AWS-speicific nature? Also, if all (one of) the constructors are from URI, we can probably skip "FromURI" suffix.
</blockquote></details>

Done.
___
*[pkg/storage/cloudimpl/cloudimpltests/aws_kms_test.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51884#-MD1Bxjr5iy71B_H3nYO:-MDFUXfa5bdJheqqXLTz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/266867053f6b13947a3d2dfe40b03b5fa4286ad4/pkg/storage/cloudimpl/cloudimpltests/aws_kms_test.go#L68)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

extra \n?
</blockquote></details>

Done.

**Comment by adityamaru** [comment]
TFTRs!

bors r+

**Pull request merged**